### DISCUSSION TITLE
Stored data de-duplication

### DISCUSSION THREAD

**Discussion by cayolblake** [juicedata/juicefs#330] [Ideas]
Hello,

I wonder if data de-duplication is on your roadmap.

Basically, if an attempt to store the same file multiple times would first check if the file chunks already exists somewhere (by using some sort of fast hashing algorithm) and if it previously existed, it would just point to the already existing file without consuming additional space and effort.

**Comment by xiaogaozi** [comment]
It's an interesting idea. We will have a discussion internally. If you have enough time, contributions are very welcome. 😀

**Comment by davies** [comment]
De-duplication is hot topic in storage, and useful for some cases, may not justify for the high cost as a general feature, so we don't put that in the roadmap.

JuiceFS does provide some features to dedup the data at application level, for example, hardlink and faster CopyFileRange (without coping the data).

**Reply by xyb** [reply]
I'm researching how to implement out-of-band deduplication using `copy_file_range` like btrfs and duperemove. I want to understand the level of support offered by JuiceFS. Does it only work with block boundaries and if so, what is the block size? Alternatively, does JuiceFS support flexible block size similar to the rolling hash generated?

**Reply by davies** [reply]
There is no limit on CopyFileRange, it work with any length from anywhere. 

Ideally, you should take aware of chunk boundary (64MiB), which means two files share the same 64MiB. Otherwise, A file may refer to partial of a slice in another file. When there are too many slices (> 5), it may trigger compaction to duplicate the data.

**Reply by xyb** [reply]
@davies Thank you for the information. I believe it would be valuable to include this in the JuiceFS documentation.

**Reply by davies** [reply]
We have update this part recently: https://juicefs.com/docs/zh/community/architecture#how-juicefs-store-files
